zcm/julia/ZCM.jl

```diff
@@ -15,320 +35,253 @@ end
 type UvSub
 end
 
-type RecvBuf
-    recv_utime ::Int64;
-    zcm        ::Ptr{Native.Zcm};
-    data       ::Ptr{UInt8};
-    data_size  ::UInt32;
+immutable RecvBuf
+    recv_utime::Int64
+    zcm::Ptr{Void}
+    # TODO: This makes the assumption that char in C is 8 bits, which is not required to be true
+    data::Ptr{UInt8}
+    data_size::UInt32
 end
 
 type EventLog
 end
 
 type EventLogEvent
-    eventnum  ::Int64;
-    timestamp ::Int64;
-    chanlen   ::Int32;
-    datalen   ::Int32;
-    channel   ::Ptr{UInt8};
-    data      ::Ptr{UInt8};
+    eventnum  ::Int64
+    timestamp ::Int64
+    chanlen   ::Int32
+    datalen   ::Int32
+    channel   ::Ptr{UInt8}
+    data      ::Ptr{UInt8}
 end
 
 end
 
-# RRR: make variable naming consistent
-# RRR: improve orgainzation and make header comments more visible
-# Exported Objects and Methods
-type RecvBuf
-    rbuf  ::Ptr{Native.RecvBuf};
-    utime ::Int64;
-    data  ::Array{UInt8};
-
-    function RecvBuf(rbuf::Ptr{Native.RecvBuf})
-        instance  = new();
-        instance.rbuf = rbuf;
-
-        if (rbuf != C_NULL)
-            loadedRbuf = unsafe_load(rbuf);
-
-            instance.utime = loadedRbuf.recv_utime;
-
-            if (loadedRbuf.data != C_NULL)
-                instance.data = unsafe_wrap(Array, loadedRbuf.data, loadedRbuf.data_size);
-            else
-                instance.data = [];
-            end
-        end
+using .Native: RecvBuf
 
-        # destruction of the underlying Native.RecvBuf pointer is handled by the C lib internally
-
-        return instance;
-    end
-end
-export RecvBuf;
-
-type __zcm_handler
-    handler::Any;
-    usr::Any;
-
-    # handler must have the function or functor prototype :
-    #   (rbuf::RecvBuf, channel::String, msg::MsgType, usr)
-    __zcm_handler(handler, usr) = new(handler, usr);
+"""
+The SubscriptionOptions type contains all of the information necessary to
+decode and handle a message on the Julia side. This is the structure which is
+passed in to zcm_subscribe, and it is thus also passed to the onresponse() 
+cfunction.
+"""
+immutable SubscriptionOptions{T <: AbstractZCMType, F <: Function}
```

Comment:
What is so bad about having a usr pointer in this object? We pretty heavily use that syntax.

Reply:
`F<:Base.Callable`, or just lose the restriction on `F` (so as to allow callable types)? Though `Base.Callable` might have been introduced in 0.6.

Reply:
@tkoolen I'll just drop the restriction, yeah.

@jbendes If you want that syntax I'm happy to restore it, but I don't think passing around raw pointers (and then holding references to their "real" identities to prevent garbage collection) is the best way. Closures already do that perfectly. Fortunately, it's easy to have ZCM.jl automatically generate the closure for you. I'll add a commit and show you what I mean.

Reply:
@tkoolen I'll just drop the restriction, yeah.

@jbendes If you want that syntax I'm happy to restore it, but I don't think passing around raw pointers (and then holding references to their "real" identities to prevent garbage collection) is the best way. Closures already do that perfectly. Fortunately, it's easy to have ZCM.jl automatically generate the closure for you. I'll add a commit and show you what I mean.

Reply:
@jbendes with https://github.com/ZeroCM/zcm/pull/165/commits/df9a7381ed81c82442e1ba806b8707be07594ad8 you can pass *any number* (including 0) of additional arguments to subscribe, and they will also be passed to the handle function after the channel and message. This should cover all cases that you were using the previous user pointer for, and does so without any unsafe_converts.